src/transformers/models/molmo/modular_molmo.py

```diff
@@ -0,0 +1,2289 @@
+            attention_mask=attention_mask,
+            cache_position=cache_position,
+            num_logits_to_keep=num_logits_to_keep,
+            **kwargs,
+        )
+
+        if cache_position[0] == 0:
+            model_inputs["pixel_values"] = pixel_values
+            model_inputs["image_token_indices"] = image_token_indices
+            model_inputs["image_masks"] = image_masks
+
+        return model_inputs
+
+
+### IMAGE PROCESSING CODE
+
+
+def make_batched_images(images) -> List[List[ImageInput]]:
+    """
+    Accepts images in list or nested list format, and makes a list of images for preprocessing.
+
+    Args:
+        images (`Union[List[List[ImageInput]], List[ImageInput], ImageInput]`):
+            The input image.
+
+    Returns:
+        list: A list of images.
+    """
+    if isinstance(images, (list, tuple)) and isinstance(images[0], (list, tuple)) and is_valid_image(images[0][0]):
+        return [img for img_list in images for img in img_list]
+
+    elif isinstance(images, (list, tuple)) and is_valid_image(images[0]):
+        return images
+
+    elif is_valid_image(images):
+        return [images]
+
+    raise ValueError(f"Could not make batched video from {images}")
+
+
+def get_resize_output_image_size(
+    image: np.ndarray,
+    size: Union[int, Tuple[int, int], List[int], Tuple[int]],
+) -> tuple:
+    original_height, original_width = get_image_size(image)
+
+    scale_y = size["height"] / original_height
+    scale_x = size["width"] / original_width
+    scale = min(scale_x, scale_y)
+
+    # Compute new dimensions
+    new_height = round(original_height * scale)
+    new_width = round(original_width * scale)
+    return {"height": new_height, "width": new_width}
+
+
+def pad_to_bounding_box(
+    image: np.ndarray, offset_height: int, offset_width: int, target_height: int, target_width: int, value: int = 0
+) -> np.ndarray:
+    """
+    Pad the input image to the target height and width using the transformers `pad` function.
+
+    Args:
+        image: The input image to be padded.
+        offset_height: The number of pixels to add to the top of the image.
+        offset_width: The number of pixels to add to the left of the image.
+        target_height: The target height of the padded image.
+        target_width: The target width of the padded image.
+        value: The constant value used for padding (default is 0).
+
+    Returns:
+        A padded image of size (target_height, target_width).
+    """
+    height, width = image.shape[:2]
+    after_padding_height = target_height - offset_height - height
+    after_padding_width = target_width - offset_width - width
+    return np.pad(
+        image,
+        [
+            (offset_height, after_padding_height),
+            (offset_width, after_padding_width),
+            (0, 0),  # don't pad on the channel dim
+        ],
+        mode="constant",
+        constant_values=value,
+    )
+
+
+class MolmoImageProcessor(BaseImageProcessor):
```

Comment:
if there is not inheritance, why is it in the modular file?

Reply:
Oh, I had an idea that we use `modular` as the only source of truth for all code. Can remove it entirely then